### PR TITLE
[DEPRECATED]  To implement gradient for Pytorch

### PR DESCRIPTION
**Pull request by iramazanli**
Fixes ##56129

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -1479,6 +1479,27 @@
...
+  variants: function
+
+- func: gradient.scalarrayarray(Tensor self, *, Scalar[] spacing, int[] dim, int edge_order=1) -> Tensor[]
+  variants: function
+
+- func: gradient.tensorarrayint(Tensor self, *, Tensor[] spacing, int? dim=None, int edge_order=1) -> Tensor[]
+  variants: function
+
+- func: gradient.tensorarray(Tensor self, *, Tensor[] spacing, int[] dim, int edge_order=1) -> Tensor[]
+  variants: function
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:1501]: Are you going to add out variants of these functions? If so, consider making it structured https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml:1501]: BTW, this is an annoyingly large number of overloads; maybe this was in the issue desc

Comment (iramazanli) [reply aten/src/ATen/native/native_functions.yaml:1501]: By the way, i think the signature ordering we have added in order to fix issue regarding {h/v/d}split methods, causing problem here. As we can see currently CI tests tells our signature matching goes to Tensor[] while we would hope to go for int[]. However after undo'ing the change i have proposed in the previous PR (https://github.com/pytorch/pytorch/pull/57346) this problem no more happens. 

What would you think is a good solution at this point ?
Should we return back to split methods / ordering function, or there is some idea that would be good to apply ?

Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml:1501]: Reverting doesn't sound right. It sounds like your problem is you want `Scalar[]` to be tested before `Tensor[]`: *so say that in the ordering*.

Comment (iramazanli) [reply aten/src/ATen/native/native_functions.yaml:1501]: Thank you for your suggestion :) and i have updated ordering accordingly. Hope it looks good now :)

Comment (iramazanli) [reply aten/src/ATen/native/native_functions.yaml:1501]: > Are you going to add out variants of these functions? If so, consider making it structured https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

Thanks for pointing the link. Currently adding methods are not planned, but will definitely use the help of this link when we are going to add it.

File: tools/autograd/gen_python_functions.py [hunk 2]
```diff
@@ -763,7 +763,8 @@ def is_arg_smaller(t1: Type, t2: Type) -> bool:
         # discussed why it is important to prioritize int/int? over int[]
```
Comment (ezyang) [inline tools/autograd/gen_python_functions.py:763]: Since there are now a lot of special cases it seems like it would be better to put the comment directly above the relevant line

File: tools/autograd/gen_python_functions.py [hunk 3]
```diff
@@ -763,7 +763,8 @@ def is_arg_smaller(t1: Type, t2: Type) -> bool:
...
         return (str(t1) == 'Scalar' and str(t2) == 'Tensor' or
                 'Dimname' in str(t1) and 'Dimname' not in str(t2) or
-                str(t1) == 'int[]' and (str(t2) == 'int' or str(t2) == 'int?'))
+                str(t1) == 'int[]' and (str(t2) == 'int' or str(t2) == 'int?') or
+                str(t1) == 'Tensor[]' and (str(t2) == 'Scalar[]'))
```
Comment (ezyang) [inline tools/autograd/gen_python_functions.py:767]: Looking at this here I see this is asymmetric with `str(t1) == 'Scalar' and str(t2) == 'Tensor'`. Why?

Comment (iramazanli) [reply tools/autograd/gen_python_functions.py:767]: The reason can be illustrated in the following example which leads to an error : 
```
>>> t = torch.tensor([[1,2,3],[4,5,6],[7,8,9]])
>>> dim = (0, 1)
>>> spacing = [0.1, 0.2]
>>> torch.gradient(t, spacing=spacing, dim=dim, edge_order=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: expected Tensor as element 0 in argument 1, but got float
```

Without prioritizing Scalar[] over Tensor[], the call goes to Tensor[] instead of Scalar[] and leads to this error.  As it has been suggested in the comment above "Reverting doesn't sound right. It sounds like your problem is you want Scalar[] to be tested before Tensor[]: so say that in the ordering." i added this new ordering method.

Do you think that, there is another alternative that i can follow ?

Thank :)

File: tools/autograd/gen_python_functions.py (hunk 2 above)
Comment (iramazanli) [reply tools/autograd/gen_python_functions.py:763]: Thanks for recommending and it is done: https://github.com/pytorch/pytorch/pull/54617/files#  
(Sorry for the confusion, there are to PR's regarding this task, for some reason and #54617 is the one we intend to merge)

File: tools/autograd/gen_python_functions.py (hunk 3 above)
Comment (ezyang) [reply tools/autograd/gen_python_functions.py:767]: There is something fishy going on here  and I'd like you to investigate (unless, @mruberry, you think this is not worth investigating during onramping). Your reasoning here is very reasonable, but it raises more questions:

1. When Tensor[] is put before Scalar[], you get a type error saying Tensor was expected but Scalar was gotten instead. But the whole point of overloads is that if you don't match the overload, you don't error, you try the next one. Why is that not happening here? (I suspect it's something to do with us only sniffing the topmost object, which is just a list and we don't know what kind of list it is)
2. Similarly, in the line above, we prioritize Tensor before Scalar. This is the "wrong" ordering. Under the reasoning you've applied above, this ought not to work, because whenever we pass a scalar, you should get a "Tensor was expected but not scalar". But empirically, this appears to work, why?

Comment (mruberry) [reply tools/autograd/gen_python_functions.py:767]: Sounds worth investigating to me ;)

We should strive to be very precise about these changes if we're pursuing them.

Comment (iramazanli) [reply tools/autograd/gen_python_functions.py:767]: I will investigate it tomorrow :)

Comment (gchanan) [reply tools/autograd/gen_python_functions.py:767]: This is good analysis, thanks.

There are a couple of strange things here for me:
1. all the other call sites to is_tensor_and_append_overloaded pass false for `throw_error` -- how are they able to do the correct thing?
2. the other list checks do the following checks:
 - is the parameter a list/sequence/etc
 - if there is a first element, does it match the type of list?

the later check works out because: 1) all empty lists should behave the same way, 2) if you have "mismatched list types" you will throw an error later, when you try to extract the parameters to C++.  This would fix your case because the the scalar list arguments just wouldn't match the TensorList version.

But torch function wants to handle all the TensorList parameters upfront, probably because you don't actually call signature-specific overrides for torch_function, otherwise you'd probably have to "push down" the torch function matching into each signature and either check if you need to call torch function (if nothing matched in the list), or to extra other overloads.  I'm also not sure if there is an ordering constraint, but that should be solvable.

So, long story short, it seems like:
1) we could make torch function handling more similar to how we handle other lists
2) we could skip the error, but we'd have to be sure TensorList was last in the list precedence; your change makes it so that ScalarList is before, but I'm unsure about IntList.  And we'll probably have more lists.

What do you think @ezyang?

Comment (ezyang) [reply tools/autograd/gen_python_functions.py:767]: Yes, I now remember why the tensor list code is this way, and it is exactly because we need to eagerly verify if any member of the list is overloaded so we can short circuit to torch function handling. When Hameer originally implemented this in the "obvious way" (don't throw an error here, but loop over everything looking for torch function, and then on extraction loop again check if you have a tensor), this regressed performance, so we added the fastpath.

I kind of agree with @gchanan that deleting the error raising logic in this case seems like it ought to work, but it is so simple I'm not sure why the code wasn't written this way to begin with. You should give it a try and see what breaks. I don't think we necessarily have to put TensorList at the end of the precedence list, if it is earlier we'll just try to match TensorList, fail, and move on to the next overload.

cc @hameerabbasi

Comment (gchanan) [reply tools/autograd/gen_python_functions.py:767]: I was thinking you want it at the end because it's potentially doing the entire traversal, but the performance in that case probably doesn't matter.

Comment (iramazanli) [reply tools/autograd/gen_python_functions.py:767]: At this point, i have tried to modify the checking process for TensorList without throwing error and it broke following test 

```
======================================================================
FAIL [0.010s]: test_cat2_cpu_float16 (__main__.TestTensorCreationCPU)
----------------------------------------------------------------------
TypeError: cat(): argument 'tensors' (position 1) must be tuple of Tensors, not list
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 729, in only_fn
    return fn(slf, device, *args, **kwargs)
  File "test_tensor_creation_ops.py", line 1158, in test_cat2
    self.assertRaisesRegex(TypeError, 'got None', lambda: torch.cat([x, None]))
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1452, in assertRaisesRegex
    return super().assertRaisesRegex(expected_exception, expected_regex, *args, **kwargs)
AssertionError: "got None" does not match "cat(): argument 'tensors' (position 1) must be tuple of Tensors, not list"
```

and my tests regarding giving TensorList lowest priority among all lists is still running and I will update the result here once testing is done.

Comment (ezyang) [reply tools/autograd/gen_python_functions.py:767]: This doesn't seem like a good reason to bin the change; this just looks like the error message wobbled which is to be expected

Comment (iramazanli) [reply tools/autograd/gen_python_functions.py:767]: I mentioned the problem regarding TensorList argument parsing in the issue :  https://github.com/pytorch/pytorch/issues/58087 and would be happy for further discussion if you're interested. At this point, gradient PR is merged as it was our for a while. Thanks for walking me through the process :)
